rollup/ccc/logger.go

```diff
@@ -0,0 +1,859 @@
+		})
+	} else {
+		l.logRawBytecode(input) // init bytecode
+	}
+
+	if !env.TxContext.IsL1MessageTx {
+		l.sigCount++
+		l.l2TxnsRlpSize += uint64(env.TxContext.TxSize)
+	}
+	l.keccakUsage += computeKeccakRows(uint64(env.TxContext.TxSize))
+	l.keccakUsage += computeKeccakRows(64) // ecrecover per txn
+}
+
+func (l *Logger) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	if err != nil {
+		return
+	}
+
+	l.evmUsage += evmUsagePerOpCode[op]
+	l.stateUsage += stateUsagePerOpCode[op](scope, depth)
+
+	getInputFn := func(inputOffset int64) func(int64, int64) []byte {
+		return func(argOffset, argLen int64) []byte {
+			// it is safe to ignore the error, with current block gas limits (10M) it is not possible to go past
+			// GetMemoryCopyPadded's internal padding limit (10M) or trigger and overflow in offset/len
+			input, _ := tracers.GetMemoryCopyPadded(scope.Memory, inputOffset+argOffset, argLen)
+			return input
+		}
+	}
+
+	switch op {
+	case vm.EXTCODECOPY:
+		l.logBytecodeAccessAt(scope.Stack.Back(0).Bytes20())
+		l.logCopy(scope.Stack.Back(3).Uint64())
+	case vm.CALLDATACOPY, vm.RETURNDATACOPY, vm.CODECOPY, vm.MCOPY, vm.CREATE, vm.CREATE2:
+		l.logCopy(scope.Stack.Back(2).Uint64())
+	case vm.SHA3:
+		l.keccakUsage += computeKeccakRows(scope.Stack.Back(1).Uint64())
+		l.logCopy(scope.Stack.Back(1).Uint64())
+	case vm.LOG0, vm.LOG1, vm.LOG2, vm.LOG3, vm.LOG4, vm.RETURN, vm.REVERT:
+		l.logCopy(scope.Stack.Back(1).Uint64())
+	case vm.DELEGATECALL, vm.STATICCALL:
+		inputOffset := int64(scope.Stack.Back(2).Uint64())
+		inputLen := int64(scope.Stack.Back(3).Uint64())
+		l.logCall(scope.Stack.Back(1).Bytes20(), uint64(inputLen), getInputFn(inputOffset))
+	case vm.CALL, vm.CALLCODE:
+		inputOffset := int64(scope.Stack.Back(3).Uint64())
+		inputLen := int64(scope.Stack.Back(4).Uint64())
+		l.logCall(scope.Stack.Back(1).Bytes20(), uint64(inputLen), getInputFn(inputOffset))
+	case vm.EXP:
+		const rowsPerExpCall = 8
+		l.expUsage += rowsPerExpCall
+	}
+}
+
+func (l *Logger) CaptureStateAfter(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	if err != nil {
+		return
+	}
+
+	switch op {
+	case vm.CREATE, vm.CREATE2:
+		l.logBytecodeAccessAt(scope.Stack.Back(0).Bytes20()) // deployed bytecode
+	}
+}
+
+func (l *Logger) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	switch typ {
+	case vm.CREATE, vm.CREATE2:
+		l.logRawBytecode(input) // init bytecode
+	}
+}
+
+func (l *Logger) CaptureExit(output []byte, gasUsed uint64, err error) {
+
+}
+
+func (l *Logger) CaptureFault(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
+
+}
+
+func (l *Logger) CaptureEnd(output []byte, gasUsed uint64, t time.Duration, err error) {
+	if l.isCreate && err != nil {
+		l.logRawBytecode(output) // deployed bytecode
```

Comment:
Isn't it a duplicate with the call in `CaptureStateAfter` and `CaptureEnter`? Or are these called for different contexts?

Reply:
`CaptureEnter` counts the init code (ie the code that generates the contract code) while `CaptureStateAfter` counts the deployed bytecode (the actual contract code).